File: Sources/Ably/ARTChannels.m
#import "ARTChannels+Private.h"

#import "ARTRest+Private.h"
#import "ARTChannel+Private.h"
#import "ARTChannelOptions.h"
#import "ARTRestChannel.h"

@interface ARTChannels() {
    __weak id<ARTChannelsDelegate> _delegate; // weak because delegates outlive their counterpart
    dispatch_queue_t _queue;
}

@end

@implementation ARTChannels

- (instancetype)initWithDelegate:(id)delegate dispatchQueue:(dispatch_queue_t)queue prefix:(NSString *)prefix {
    if (self = [super init]) {
        _queue = queue;
        _channels = [[NSMutableDictionary alloc] init];
        _delegate = delegate;
        _prefix = prefix;
    }
    return self;
}

- (id<NSFastEnumeration>)copyIntoIteratorWithMapper:(id (^)(id))mapper {
    __block id<NSFastEnumeration>ret;
dispatch_sync(_queue, ^{
    NSMutableArray *channels = [[NSMutableArray alloc] init];
    for (id internalChannel in [self getNosyncIterable]) {
        [channels addObject:mapper(internalChannel)];
    }
    ret = [channels objectEnumerator];
});
    return ret;
}

- (id<NSFastEnumeration>)getNosyncIterable {
    return [_channels objectEnumerator];
}

- (BOOL)exists:(NSString *)name {
    __block BOOL ret;
dispatch_sync(_queue, ^{
    ret = [self _exists:name];
});
    return ret;
}

- (BOOL)_exists:(NSString *)name {
    return self->_channels[[self addPrefix:name]] != nil;
}

- (id)get:(NSString *)name {
    return [self getChannel:[self addPrefix:name] options:nil];
}

- (id)get:(NSString *)name options:(ARTChannelOptions *)options {
    return [self getChannel:[self addPrefix:name] options:options];
}

- (void)release:(NSString *)name {
dispatch_sync(_queue, ^{
    [self _release:name];
});
}

- (void)_release:(NSString *)name {
    [self->_channels removeObjectForKey:[self addPrefix:name]];
}

- (ARTRestChannel *)getChannel:(NSString *)name options:(ARTChannelOptions *)options {
    __block ARTRestChannel *channel;
dispatch_sync(_queue, ^{
    channel = [self _getChannel:name options:options addPrefix:true];
});
    return channel;
}

- (ARTChannel *)_getChannel:(NSString *)name options:(ARTChannelOptions *)options addPrefix:(BOOL)addPrefix {
    if (addPrefix) {
        name = [self addPrefix:name];
    }
    ARTChannel *channel = [self _get:name];
    if (!channel) {
        channel = [_delegate makeChannel:name options:options];
        [self->_channels setObject:channel forKey:name];
    } else if (options) {
        [channel setOptions_nosync:options];
    }
    return channel;
}

- (id)_get:(NSString *)name {
    return self->_channels[name];
}

- (NSString *)addPrefix:(NSString *)name {
    if (_prefix) {
        if (![name hasPrefix:_prefix]) {
            return [NSString stringWithFormat:@"%@-%@", _prefix, name];
        }
    }
    return name;
}

@end

File: Sources/AblySwift/ARTChannels.swift
import Foundation

// swift-migration: original location ARTChannels.h, line 11 and ARTChannels.m, line 15
internal class ARTChannels<ChannelType>: NSObject where ChannelType: ARTChannel {
    private weak var delegate: ARTChannelsDelegate? // weak because delegates outlive their counterpart
    private let queue: DispatchQueue
    
    // swift-migration: original location ARTChannels+Private.h, line 16 and ARTChannels.m, line 20
    internal let channels: NSMutableDictionary
    
    // swift-migration: original location ARTChannels+Private.h, line 18
    internal let prefix: String?
    
    // swift-migration: original location ARTChannels+Private.h, line 27 and ARTChannels.m, line 17
    internal init(delegate: ARTChannelsDelegate, dispatchQueue: DispatchQueue, prefix: String?) {
        self.queue = dispatchQueue
        self.channels = NSMutableDictionary()
        self.delegate = delegate
        self.prefix = prefix
        super.init()
    }
    
    // swift-migration: original location ARTChannels.h, line 50 and ARTChannels.m, line 27
    internal func copyIntoIterator(withMapper mapper: @escaping (ChannelType) -> Any) -> NSFastEnumeration {
        var ret: NSFastEnumeration!
        queue.sync {
            let channelsArray = NSMutableArray()
            let enumerator = getNosyncIterable as! NSEnumerator
            while let nextObj = enumerator.nextObject() {
                channelsArray.add(mapper(nextObj as! ChannelType))
            }
            ret = channelsArray.objectEnumerator()
        }
        return ret
    }
    
    // swift-migration: original location ARTChannels+Private.h, line 17 and ARTChannels.m, line 39
    internal var getNosyncIterable: NSFastEnumeration {
        return channels.objectEnumerator()
    }
    
    // swift-migration: original location ARTChannels.h, line 20 and ARTChannels.m, line 43
    internal func exists(_ name: String) -> Bool {
        var ret: Bool = false
        queue.sync {
            ret = _exists(name)
        }
        return ret
    }
    
    // swift-migration: original location ARTChannels+Private.h, line 22 and ARTChannels.m, line 51
    internal func _exists(_ name: String) -> Bool {
        return channels[addPrefix(name)] != nil
    }
    
    // swift-migration: original location ARTChannels.h, line 29 and ARTChannels.m, line 55
    internal func get(_ name: String) -> ChannelType {
        return getChannel(addPrefix(name), options: nil) as! ChannelType
    }
    
    // swift-migration: original location ARTChannels.h, line 39 and ARTChannels.m, line 59
    internal func get(_ name: String, options: ARTChannelOptions) -> ChannelType {
        return getChannel(addPrefix(name), options: options) as! ChannelType
    }
    
    // swift-migration: original location ARTChannels.h, line 47 and ARTChannels.m, line 63
    internal func release(_ name: String) {
        queue.sync {
            _release(name)
        }
    }
    
    // swift-migration: original location ARTChannels+Private.h, line 25 and ARTChannels.m, line 69
    internal func _release(_ name: String) {
        channels.removeObject(forKey: addPrefix(name))
    }
    
    // swift-migration: original location ARTChannels.m, line 73
    internal func getChannel(_ name: String, options: ARTChannelOptions?) -> ARTChannel {
        var channel: ARTChannel!
        queue.sync {
            channel = _getChannel(name, options: options, addPrefix: true)
        }
        return channel
    }
    
    // swift-migration: original location ARTChannels+Private.h, line 24 and ARTChannels.m, line 81
    internal func _getChannel(_ name: String, options: ARTChannelOptions?, addPrefix: Bool) -> ARTChannel {
        var channelName = name
        if addPrefix {
            channelName = self.addPrefix(name)
        }
        var channel = _get(channelName)
        if channel == nil {
            channel = delegate?.makeChannel(channelName, options: options)
            if let channel = channel {
                channels.setObject(channel, forKey: channelName as NSString)
            }
        } else if let options = options {
            channel?.setOptions_nosync(options)
        }
        return channel!
    }
    
    // swift-migration: original location ARTChannels+Private.h, line 23 and ARTChannels.m, line 95
    internal func _get(_ name: String) -> ARTChannel? {
        return channels[name] as? ARTChannel
    }
    
    // swift-migration: original location ARTChannels+Private.h, line 20 and ARTChannels.m, line 99
    internal func addPrefix(_ name: String) -> String {
        if let prefix = self.prefix {
            if !name.hasPrefix(prefix) {
                return "\(prefix)-\(name)"
            }
        }
        return name
    }
}